firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+                     "All integrals must share the same domain of integration.")
+        return domains[0]
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        `{domain:{integral_type: subdomain_data}}`."""
+
+        # Scalar case
+        if self.rank == 0:
+            return None
+        integration_domains = self.ufl_domains()
+        integrals = self._integrals
+
+        # Initializing subdomain_data
+        subdomain_data = {}
+        for domain in integration_domains:
+            subdomain_data[domain] = {}
+
+        for integral in integrals:
+            domain = integral.ufl_domain()
+            it_type = integral.integral_type()
+            subdata = integral.subdomain_data()
+
+            data = subdomain_data[domain].get(it_type)
+            if data is None:
+                subdomain_data[domain][it_type] = subdata
+            elif subdata is not None:
+                slate_assert(data.ufl_id() == subdata.ufl_id(),
+                             "Integrals in the tensor must have the same subdomain_data objects.")
+
+        return subdomain_data
+
+    # Signature and hash methods
+    def _compute_renumbering(self):
+        """Renumbers coefficients by including integration domains."""
+
+        dn = self._domain_numering
+        cn = self._coefficient_numbering
+        renumbering = {}
+        renumbering.update(dn)
+        renumbering.update(cn)
+
+        n = len(dn)
+        for c in cn:
+            d = c.ufl_domain()
+            if d is not None and d not in renumbering:
+                renumbering[d] = n
+                n += 1
+
+        return renumbering
+
+    def _compute_signature(self, renumbering):
+        """Computes the signature of the tensor from the integrals."""
+        # TODO: Is there a better/cheaper way of doing this?!
+
+        integrals = self._integrals
+        integrands = [integral.integrand() for integral in integrals]
+
+        terminal_hashdata = compute_terminal_hashdata(integrands, renumbering)
+
+        hashdata = []
+        for integral in integrals:
+            integrand_hashdata = compute_expression_hashdata(integral.integrand(),
+                                                             terminal_hashdata)
+            domain_hashdata = integral.ufl_domain()._ufl_signature_data_(renumbering)
+
+            integral_hashdata = (integrand_hashdata, domain_hashdata,
+                                 integral.integral_type(),
+                                 integral.subdomain_id(),
+                                 canonicalize_metadata(integral.metadata()), )
+            hashdata.append(integral_hashdata)
+
+        return hashlib.sha512(str(hashdata).encode('utf-8')).hexdigest()
+
+    def signature(self):
+        """Signature for use in caching. Reproduced from ufl."""
+
+        if self._signature is None:
+            self._signature = self._compute_signature(self._compute_renumbering())
+
+        return self._signature
```

Comment:
I like the idea of SLATE being semi-autonomous, and hence the base class `Tensor` not inheriting from `ufl.Form`. (Because it isn't a form!) All the methods reproduced here were added as I discovered that I needed them. I think your suggestion removes a lot of unnecessary code duplication